tests/core/math/test_projection.h

```diff
@@ -0,0 +1,664 @@
+		Plane expected = Plane(Vector3(0, 0, 1), 1);
+		Plane actual = Projection().get_projection_plane(Projection::Planes::PLANE_FAR);
+
+		CHECK(expected == actual);
+	}
+
+	SUBCASE("PLANE_LEFT") {
+		Plane expected = Plane(Vector3(-1, 0, 0), 1);
+		Plane actual = Projection().get_projection_plane(Projection::Planes::PLANE_LEFT);
+
+		CHECK(expected == actual);
+	}
+
+	SUBCASE("PLANE_TOP") {
+		Plane expected = Plane(Vector3(0, 1, 0), 1);
+		Plane actual = Projection().get_projection_plane(Projection::Planes::PLANE_TOP);
+
+		CHECK(expected == actual);
+	}
+
+	SUBCASE("PLANE_RIGHT") {
+		Plane expected = Plane(Vector3(1, 0, 0), 1);
+		Plane actual = Projection().get_projection_plane(Projection::Planes::PLANE_RIGHT);
+
+		CHECK(expected == actual);
+	}
+
+	SUBCASE("PLANE_BOTTOM") {
+		Plane expected = Plane(Vector3(0, -1, 0), 1);
+		Plane actual = Projection().get_projection_plane(Projection::Planes::PLANE_BOTTOM);
+
+		CHECK(expected == actual);
+	}
+}
+
+TEST_CASE("[Projection] get_viewport_half_extents ") {
+	SUBCASE("get_viewport_half_extents of a generic matrix") {
+		Vector2 expected(0.189962f, 0.867947f);
+		Vector2 actual = Projection(
+				Vector4(-29.32f, 43.39f, -94.85f, -89.95f),
+				Vector4(55.01f, -55.50f, 87.92f, 36.67f),
+				Vector4(48.63f, 19.80f, 75.48f, 44.12f),
+				Vector4(-2.16f, 70.96f, -59.63f, 23.87f))
+								 .get_viewport_half_extents();
+
+		CHECK(expected.x == doctest::Approx(actual.x));
+		CHECK(expected.y == doctest::Approx(actual.y));
+	}
+
+	SUBCASE("get_viewport_half_extents of null matrix") {
+		Vector2 expected(0.0f, 0.0f);
+		Vector2 actual = Projection(
+				Vector4(0, 0, 0, 0),
+				Vector4(0, 0, 0, 0),
+				Vector4(0, 0, 0, 0),
+				Vector4(0, 0, 0, 0))
+								 .get_viewport_half_extents();
+
+		CHECK(expected.is_equal_approx(actual));
+	}
+}
+
+TEST_CASE("[Projection] get_z_far") {
+	float expected = 1;
+	float actual = Projection().get_z_far();
+
+	CHECK(expected == actual);
+}
+
+TEST_CASE("[Projection] get_z_near") {
+	float expected = -1;
+	float actual = Projection().get_z_near();
+
+	CHECK(expected == actual);
+}
+
+TEST_CASE("[Projection] inverse") {
+	SUBCASE("The inverse of the identity is the identity") {
+		Projection expected;
+		Projection actual = Projection().inverse();
+
+		for (int i = 0; i < 4; i++) {
+			for (int j = 0; j < 4; j++) {
+				CHECK(expected[i][j] == doctest::Approx(actual[i][j]));
+			}
+		}
+	}
+
+	SUBCASE("Calculate the inverse of a matrix") {
+		Projection expected(
+				Vector4(0.0145519, -0.0304995, -0.00444437, 0.0161966),
+				Vector4(0.00190803, 0.01359, 0.0164218, -0.00228562),
+				Vector4(-0.00297949, 0.0167925, 0.0164924, -0.015914),
+				Vector4(-0.0117984, -0.00121019, -0.00802058, 0.0103989));
+		Projection actual = Projection(
+				Vector4(-29.32f, 43.39f, -94.85f, -89.95f),
+				Vector4(-55.01f, 55.50f, -87.92f, -36.67f),
+				Vector4(48.63f, 19.80f, 75.48f, 44.12f),
+				Vector4(-2.16f, 70.96f, -59.63f, 23.87f))
+									.inverse();
```

Comment:
You are mixing double and single precision constants through your code, is there a reason for this? I think this might cause problems with double precision builds

Reply:
No, there is no specific reason. I was just confused about that f at the end of floating numbers. I am now aware of what is it for and I proceeded to remove it as it is not needed in this case.